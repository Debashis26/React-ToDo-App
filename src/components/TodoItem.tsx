import React, { useState } from "react";
import "./TodoItem.css";
import { CSSTransition, TransitionGroup } from "react-transition-group";
import { ITodo, RemoveTodo, UpdateTodo, ToggleTodo } from "../types";

interface TodoItemProps {
  removeTodo: RemoveTodo;
  updateTodo: UpdateTodo;
  toggleTodo: ToggleTodo;
  todo: ITodo;
}

const TodoItem: React.FC<TodoItemProps> = ({
  removeTodo,
  todo,
  updateTodo,
  toggleTodo,
}) => {
  const [mode, setMode] = useState(false);
  const [updateTask, setUpdateTask] = useState(todo.task);

  const handleUpdate = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    if (!updateTask || /^\s*$/.test(updateTask)) return;
    updateTodo(todo.id, updateTask);
    setMode(!mode);
  };

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setUpdateTask(e.target.value);
  };

  let result: JSX.Element;
  if (mode) {
    result = (
      <CSSTransition key="editing" timeout={500} classNames="form">
        <form className="Todo-edit-form" onSubmit={handleUpdate}>
          <input
            type="text"
            value={updateTask}
            name="task"
            onChange={handleChange}
          />
          <button>Save</button>
        </form>
      </CSSTransition>
    );
  } else {
    result = (
      <CSSTransition key="normal" timeout={500} classNames="task-text">
        <li className="Todo-task" onClick={() => toggleTodo(todo.id)}>
          {todo.task}
        </li>
      </CSSTransition>
    );
  }

  return (
    <TransitionGroup className={todo.completed ? "Todo completed" : "Todo"}>
      {result}
      <CSSTransition timeout={500}>
        <div className="Todo-buttons">
          <button onClick={() => setMode(!mode)}>
            <i className="fas fa-pen" />
          </button>
          <button onClick={() => removeTodo(todo.id)}>
            <i className="fas fa-trash" />
          </button>
        </div>
      </CSSTransition>
    </TransitionGroup>
  );
};
export default TodoItem;
