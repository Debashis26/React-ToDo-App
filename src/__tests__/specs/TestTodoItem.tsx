import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import TodoItem from '../../components/TodoItem';

const mockTodo = {
  id: 1,
  task: 'Example Task',
  completed: false,
};

const mockRemoveTodo = jest.fn();
const mockUpdateTodo = jest.fn();
const mockToggleTodo = jest.fn();

const setup = (customProps = {}) => {
  const props = {
    removeTodo: mockRemoveTodo,
    updateTodo: mockUpdateTodo,
    toggleTodo: mockToggleTodo,
    todo: { ...mockTodo, ...customProps },
  };

  const utils = render(<TodoItem {...props} />);
  return { ...utils, props };
};

describe('Testing TodoItem Component', () => {
  it('Should renders correctly', () => {
    const { container } = setup();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('Should renders task text when not in edit mode', () => {
    const { getByText } = setup();    
    expect(getByText(mockTodo.task)).toBeInTheDocument();
  });

  it('Should toggles edit mode when edit button is clicked', () => {
    const { getByTestId } = setup();
    fireEvent.click(getByTestId('edit'));
    expect(mockUpdateTodo).not.toHaveBeenCalled(); // Ensure updateTodo is not called immediately
  });

  it('Should calls updateTodo when form is submitted with a new task', () => {
    const { getByTestId, getByText } = setup();
    fireEvent.click(getByTestId('edit'));

    const newTask = 'New Task';
    fireEvent.change(screen.getByRole('textbox'), { target: { value: newTask } });
    fireEvent.click(getByText('Save'));

    expect(mockUpdateTodo).toHaveBeenCalledWith(mockTodo.id, newTask);
  });

  it('Should calls removeTodo when delete button is clicked', () => {
    const { getByTestId } = setup();
    fireEvent.click(getByTestId('delete'));
    expect(mockRemoveTodo).toHaveBeenCalledWith(mockTodo.id);
  });

  it('Should mark over the test to denote the task is completed when task is clicked', () => {
    const { getByText } = setup();
    fireEvent.click(getByText(mockTodo.task));
    expect(mockToggleTodo).toHaveBeenCalledWith(mockTodo.id);
  });
});
