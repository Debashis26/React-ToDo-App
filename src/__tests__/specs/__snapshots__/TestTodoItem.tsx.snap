// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing TodoItem Component Should renders correctly 1`] = `
<div
  class="Todo"
>
  <li
    class="Todo-task"
  >
    Example Task
  </li>
  <div
    class="Todo-buttons"
  >
    <button>
      <i
        class="fas fa-pen"
        data-testid="edit"
      />
    </button>
    <button>
      <i
        class="fas fa-trash"
        data-testid="delete"
      />
    </button>
  </div>
</div>
`;

exports[`TodoItem renders correctly 1`] = `
<div
  class="Todo"
>
  <li
    class="Todo-task"
  >
    Example Task
  </li>
  <div
    class="Todo-buttons"
  >
    <button>
      <i
        class="fas fa-pen"
        data-testid="edit"
      />
    </button>
    <button>
      <i
        class="fas fa-trash"
        data-testid="delete"
      />
    </button>
  </div>
</div>
`;
